### ISSUE TITLE
[bug] TestPackage test_type = "explicit" not works with --lockfile

### ISSUE THREAD

**Issue by hekrdla** [conan-io/conan#14833]
### Environment details

* Operating System+version: windows 10
* Compiler+version: msvc v142
* Conan version: 1.61.0
* Python version: 3.10.10

### Steps to reproduce

conan lock create ... --lockfile-out xyz.lock
conan create ... --lockfile xyz.lock

issue found with protobuf package
`ERROR: Something failed while testing 'protobuf/3.5.1@CMP/stable' test_package after it was built using the lockfile. Please report this error: 'protobuf'`

## Exception info
[create.py](https://github.com/conan-io/conan/blob/release/1.61/conans/client/cmd/create.py#L67) - added by me
```
Traceback (most recent call last):
  File "Python310\site-packages\conans\client\cmd\create.py", line 64, in create
    install_build_and_test(app, test_conanfile_path, ref, graph_info, remotes,
  File "Python310\site-packages\conans\client\cmd\test.py", line 35, in install_build_and_test
    install_folder, conanfile = deps_install(app=app,
  File "Python310\site-packages\conans\client\manager.py", line 58, in deps_install
    deps_graph = graph_manager.load_graph(ref_or_path, create_reference, graph_info, build_modes,
  File "Python310\site-packages\conans\client\graph\graph_manager.py", line 124, in load_graph
    root_node = self._load_root_node(reference, create_reference, profile_host, graph_lock,
  File "Python310\site-packages\conans\client\graph\graph_manager.py", line 166, in _load_root_node
    ret = self._load_root_test_package(path, create_reference, graph_lock, profile_host,
  File "Python310\site-packages\conans\client\graph\graph_manager.py", line 284, in _load_root_test_package
    graph_lock.find_require_and_lock(create_reference, conanfile)
  File "Python310\site-packages\conans\model\graph_lock.py", line 667, in find_require_and_lock
    conanfile.requires[reference.name].lock(locked_ref, node_id)
KeyError: 'protobuf'
```

## Error
https://github.com/conan-io/conan/blob/release/1.61/conans/model/graph_lock.py#L665

## Issue root cause:

With `test_type = "requires", "build_requires"` graph_manager init loaded test_package conanfile with tested reference before graph_lock `find_require_and_lock`.

With  `test_type = "explicit"`, graph_lock  `find_require_and_lock` runs before test_package `def requirements(self)` or `def build_requirements(self)`, so conanfile.requires is not loaded yet. (verified with exceptions in `def requirements(self)` and `def build_requirements(self)`)

**Comment by memsharded** [comment]
Hi @hekrdla 

Thanks for your report, and for all the details. It really helped to reproduce the issue.

I have managed to reproduce it with a simple test:
```python
def test_error_test_explicit():
    # https://github.com/conan-io/conan/issues/14833
    client = TestClient()
    client.save({"conanfile.py": GenConanfile("pkg", "0.1"),
                 "test_package/conanfile.py": GenConanfile().with_test("pass")
                                              .with_class_attribute("test_type = 'explicit'")})
    client.run("lock create conanfile.py --lockfile-out=my.lock")
    client.run("create . --lockfile=my.lock")
```

The good news is that this is already solved in Conan 2.0, it works without problems.

In Conan 1.X I am investigating, but the fix seems challenging. Fortunately, there is a reasonable and easy workaround, deactivating the ``test_package`` in the last command it works, so ``conan create . --lockfile=my.lock -tf=None`` works fine.

**Comment by memsharded** [comment]
I have been investigating this, and so far I can't find a viable fix for 1.X that doesn't have risk of breaking other existing users (lockfiles implementation in 1.X is very complex and fragile, they were massively improved and simplified in 2.0). I'd recommend to try to move forward with the ``-tf=None`` workaround until upgrading to 2.0.

**Comment by memsharded** [comment]
I am adding test in https://github.com/conan-io/conan/pull/14851, for 2.0, that proves that this is working fine in 2.0, that will propose to close the issue. The risk of breaking in 1.X seems too high, the recommendation of using ``-t=None`` in 1.X would still be the recommended workaround.

**Comment by memsharded** [comment]
https://github.com/conan-io/conan/pull/14851 has closed this issue as solved in 2.0, but please let us know if you have any questions regarding the workarounds for 1.X above.
